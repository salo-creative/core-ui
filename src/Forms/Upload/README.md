# Select

This is an Upload component that can be used as a `<input type='file'>` html element. It works well with `useForm`.

## Usage

Install

```bash
yarn add @salo/core-ui
```

Include

```javascript
import Upload from '@salo/core-ui/Forms/Upload';
```

Implement as follows

```javascript
<Upload 
  name='readme'
  accepts='image/*'
/>
```

Uploaders often have drastically different appearances depending on the client and context. For this reason there are 3 ways to customise the look of this upload component.

1. Pass no children like the example above and it'll use the default Button component.
2. Pass children and it'll wrap them with the label to trigger the file picker dialog. The main caveat here is that you can't nest `<button />`.

```javascript
<Upload 
  name='readme'
  accepts='image/*'
>
  Upload a file
</Upload>
```

3. And finally you can use render props with it to completely customise the look and behaviour of the trigger.

```javascript
<Upload
  name='readme'
  accepts='image/*'
>
  { ({ inputRef, disabled }) => (
    <React.Fragment>
      <P>Upload a file (pdf, docx, jpg, png)</P>
      <button
        htmlFor='readme'
        disabled={ disabled }
        onClick={ () => inputRef.current.click() }>
          Choose file
      </button>
    </React.Fragment>
   ) }
</Upload>
```

### accepts

Another conveniance features is `accepts`. This either takes an array like `['images']` or a string value like `image/png`. This allows you to think about the types of documents allowed rather than individual mime types but if you need the granularity then it's there.

```javascript
<Upload
  name='readme'
  accepts={ ['images', 'documents'] }
/>
```