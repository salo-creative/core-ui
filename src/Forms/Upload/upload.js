/* eslint-disable no-param-reassign */
import React from 'react';
import PropTypes from 'prop-types';
import { isArray } from 'lodash';

// COMPONENTS
import { InputWrapper, HiddenInput } from './upload.styles';
import ErrorText from '../components/ErrorText';
import HelperText from '../components/HelperText';
import Label from '../components/Label';

// HELPERS
import Button from '../../Molecules/Button';

export const mimeTypes = {
  documents: ['application/pdf', '.doc', '.docx', 'application/msword', 'application/vnd.openxmlformats-officedocument.wordprocessingml.document'],
  pdf: ['application/pdf'],
  images: ['image/png', 'image/jpeg', 'image/jpg'],
  data: ['text/csv']
};

function getAcceptedTypes(accept) {
  if (isArray(accept)) {
    return accept.reduce((accum, type) => {
      if (type === 'documents') {
        return accum += `,${ mimeTypes.documents.join() }`;
      } if (type === 'images') {
        return accum += `,${ mimeTypes.images.join() }`;
      } if (type === 'data') {
        return accum += `,${ mimeTypes.data.join() }`;
      }
      return accum;
    }, '');
  }
  return accept;
}

const Upload = (props) => {
  const {
    accept,
    capture,
    className,
    children,
    disabled,
    error,
    errorMessage,
    files,
    fontSize,
    helperText,
    label,
    margin,
    multiple,
    name,
    onChange,
    placeholder,
    placeholderSelected,
    required,
    size,
    showLabel
  } = props;

  const inputRef = React.useRef(null);
  const [selectedFiles, setSelectedFiles] = React.useState([]);
  const accepted = getAcceptedTypes(accept);

  const reset = () => {
    setSelectedFiles([]);
  };

  const renderTrigger = () => {
    if (typeof children === 'function') {
      return children({
        ...props,
        reset,
        inputRef,
        selectedFiles
      });
    }
    
    if (children) {
      return (
        <label htmlFor={ name }>
          { children }
        </label>
      );
    }
    
    return (
      <Button
        type='label'
        htmlFor={ name }
        id={ `${ name }_label` }
        error={ error }
        label={ label }
        name={ name }
        disabled={ disabled }
        required={ required }
        size={ size }
        className='salo-upload__label'
      >
        { !selectedFiles ? placeholder : placeholderSelected.replace('{{amount}}', selectedFiles.length) }
      </Button>
    );
  };

  return (
    <InputWrapper margin={ margin } className={ `salo-upload ${ className }` }>
      { showLabel && (
        <Label
          error={ error }
          label={ label }
          name={ name }
          required={ required }
          size={ size }
          className='salo-input__label'
        />
      ) }
      { renderTrigger() }
      <HiddenInput
        type='file'
        ref={ inputRef }
        aria-required={ required.toString() }
        aria-invalid={ error.toString() }
        accept={ accepted }
        capture={ capture }
        className='salo-upload__field'
        disabled={ disabled }
        files={ files }
        fontSize={ fontSize }
        id={ name }
        multiple={ multiple }
        name={ name }
        onChange={ (e) => {
          setSelectedFiles(e.target.files);
          onChange({
            e,
            value: multiple ? e.target.files : e.target.files[0]
          });
        } }
      />
      <ErrorText
        className='salo-upload__error'
        disabled={ disabled }
        error={ error }
        errorMessage={ errorMessage }
        size={ size }
      />
      <HelperText
        className='salo-upload__helper'
        disabled={ disabled }
        error={ error }
        helperText={ helperText }
        size={ size }
      />
    </InputWrapper>
  );
};

Upload.defaultProps = {
  accept: '*',
  capture: null,
  className: '',
  children: null,
  disabled: false,
  error: false,
  errorMessage: 'Field invalid',
  files: null,
  fontSize: '1.4rem',
  helperText: '',
  label: '',
  margin: '0 0 2rem',
  multiple: false,
  onChange: () => null,
  placeholder: 'Choose file',
  placeholderSelected: 'Chosen {{amount}} file(s)',
  required: false,
  showLabel: true,
  size: 'M'
};

Upload.propTypes = {
  accept: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),
  capture: PropTypes.string,
  className: PropTypes.string,
  children: PropTypes.any,
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  files: PropTypes.any,
  fontSize: PropTypes.string,
  helperText: PropTypes.string,
  label: PropTypes.string,
  margin: PropTypes.string,
  multiple: PropTypes.bool,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  placeholder: PropTypes.string,
  placeholderSelected: PropTypes.string,
  required: PropTypes.bool,
  showLabel: PropTypes.bool,
  size: PropTypes.oneOf(['L', 'M'])
};

export default Upload;