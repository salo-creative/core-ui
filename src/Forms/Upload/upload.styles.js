import styled from 'styled-components';
import { transparentize } from 'polished';

export const InputWrapper = styled.div`
  width: 100%;
  flex-direction: column;
  display: flex;
  margin: ${ ({ margin }) => margin };
`;

export const FieldWrapper = styled.div`
  position: relative;
`;

export const Trigger = styled.div`
  position: relative;

`;

export const HiddenInput = styled.input`
  display: none;
`;

export const Field = styled.input`
  margin: 0;
  padding: ${ ({ padding }) => padding };
  height: ${ ({ height }) => height };
  background-color: ${ ({ background, error, theme }) => (error ? transparentize(0.7, theme.error) : background) };
  transition: border 0.2s linear;
  width: 100%;
  font-size: ${ ({ fontSize }) => fontSize };

  &:focus,
  &:active {
    border-color: ${ ({ theme }) => theme.primary }
  }

  &[disabled] {
    cursor:  not-allowed;
    opacity: 0.75;
  }
`;