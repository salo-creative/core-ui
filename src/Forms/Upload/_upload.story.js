import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, text, select } from '@storybook/addon-knobs';
import { RenderWithProps, Store } from '@jamesbliss/storybook-state';

// load tests
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// FEATURED COMPONENT
import { Upload } from '../../index';

// README //
import README from './README.md';

// ADDITIONAL COMPONENTS
import P from '../../Typography/P';

// Start of story logic
const stories = storiesOf('Forms | Upload', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withTests({ results }));
stories.addParameters({ jest: ['upload'] });

stories.add(
  'Basic',
  (() => {
    // STORE
    const store = new Store({ value: '' });
    // KNOBS
    const label = text('Label', 'Input label');
    const helperText = text('Helper text', 'Helper text');
    const fontSize = text('Font size', '1.4rem');
    const error = boolean('Show error state', false);
    const required = boolean('Required field', false);
    const disabled = boolean('Show disabled state', false);
    const accept = select('accept', {
      all: '*',
      documents: ['documents'],
      images: ['images'],
      data: ['data'],
      imagesAndData: ['images', 'data']
    });
    return (
      <RenderWithProps store={ store }>
        <Upload
          name='story'
          disabled={ disabled }
          error={ error }
          fontSize={ fontSize }
          label={ label }
          required={ required }
          helperText={ helperText }
          accept={ accept }
          onChange={ ({ e }) => store.set({ value: e.target.files[0] }) }
        />
      </RenderWithProps>
    );
  }),
  {
    info: { propTablesExclude: [RenderWithProps] },
    notes: README
  }
);

stories.add(
  'Custom',
  (() => {
    // STORE
    const store = new Store({ value: '' });
    // KNOBS
    const label = text('Label', 'Input label');
    const helperText = text('Helper text', 'Helper text');
    const fontSize = text('Font size', '1.4rem');
    const error = boolean('Show error state', false);
    const required = boolean('Required field', false);
    const disabled = boolean('Show disabled state', false);
    const accept = select('accept', {
      all: '*',
      documents: ['documents'],
      images: ['images'],
      data: ['data'],
      imagesAndData: ['images', 'data']
    });
    return (
      <RenderWithProps store={ store }>
        <Upload
          name='story'
          disabled={ disabled }
          error={ error }
          fontSize={ fontSize }
          label={ label }
          required={ required }
          showLabel={ false }
          helperText={ helperText }
          accept={ accept }
          onChange={ ({ e }) => store.set({ value: e.target.files[0] }) }
        >
          <P>Upload a file (pdf, docx, jpg, png)</P>
        </Upload>
      </RenderWithProps>
    );
  }),
  {
    info: { propTablesExclude: [RenderWithProps] },
    notes: README
  }
);

stories.add(
  'Custom props',
  (() => {
    // STORE
    const store = new Store({ value: '' });
    // KNOBS
    const label = text('Label', 'Input label');
    const helperText = text('Helper text', 'Helper text');
    const fontSize = text('Font size', '1.4rem');
    const error = boolean('Show error state', false);
    const required = boolean('Required field', false);
    const disabled = boolean('Show disabled state', false);
    const accept = select('accept', {
      all: '*',
      documents: ['documents'],
      images: ['images'],
      data: ['data'],
      imagesAndData: ['images', 'data']
    });
    return (
      <RenderWithProps store={ store }>
        <Upload
          name='story'
          disabled={ disabled }
          error={ error }
          fontSize={ fontSize }
          label={ label }
          required={ required }
          helperText={ helperText }
          accept={ accept }
          onChange={ ({ e }) => store.set({ value: e.target.files[0] }) }
        >
          { ({ inputRef }) => {
            return (
              <React.Fragment>
                <P>Upload a file (pdf, docx, jpg, png)</P>
                <button
                  type='button'
                  htmlFor='story'
                  disabled={ disabled }
                  onClick={ () => inputRef.current.click() }
                >
                    Choose file
                </button>
              </React.Fragment>
            );
          } }
        </Upload>
      </RenderWithProps>
    );
  }),
  {
    info: { propTablesExclude: [RenderWithProps] },
    notes: README
  }
);