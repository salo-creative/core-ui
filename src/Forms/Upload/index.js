import Upload, { mimeTypes } from './upload';

export { mimeTypes };
export default Upload;