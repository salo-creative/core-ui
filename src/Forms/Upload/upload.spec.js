import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import Upload from './index';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <Upload />
  );
  expect(container.firstChild).toMatchSnapshot();
});