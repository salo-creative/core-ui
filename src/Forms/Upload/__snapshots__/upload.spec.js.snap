// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c0 {
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0 0 2rem;
}

.c2 {
  display: none;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  background-color: #00ACED;
  border: 1px solid #00ACED;
  color: #fff;
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
  position: relative;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: uppercase;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  border-radius: 0.4rem;
  font-size: 1.4rem;
  height: 4rem;
  padding: 0 1.5rem;
  width: auto;
  box-shadow: 2px 2px 2px 0px rgba(0,0,0,0.2);
}

.c1 path {
  -webkit-transition: fill 0.3s linear;
  transition: fill 0.3s linear;
  fill: #fff !important;
}

.c1.disabled,
.c1:disabled {
  background-color: #00ACED !important;
  border: 1px solid #00ACED;
  color: #fff;
  opacity: 0.6;
  pointer-events: auto;
  cursor: not-allowed;
}

.c1.disabled path,
.c1:disabled path {
  fill: #fff !important;
}

.c1:focus {
  outline: none;
}

.c1:hover:not([disabled]) {
  border: 1px solid #0074a1;
  background: #0074a1;
}

.c1.active,
.c1:hover:not([disabled]) {
  color: #fff;
}

.c1.active path,
.c1:hover:not([disabled]) path {
  fill: #fff !important;
}

<div
  class="c0 salo-upload "
>
  <label
    class="c1 salo-upload__label "
    height="4rem"
    id="undefined_label"
    label=""
    loading="false"
    radius="true"
  >
    <canvas
      height="0"
      style="border-radius: inherit; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
      width="0"
    />
    Chosen 0 file(s)
  </label>
  <input
    accept="*"
    aria-invalid="false"
    aria-required="false"
    class="c2 salo-upload__field"
    font-size="1.4rem"
    type="file"
  />
</div>
`;
